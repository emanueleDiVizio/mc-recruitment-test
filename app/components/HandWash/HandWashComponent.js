import React, { useState, useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { View, Text, Button } from 'react-native'
import useInterval from '../../hooks/useInterval'
import PropTypes from 'prop-types'

import handWash from '../../state/handWashSlice'

const HandWashComponent = ({ defaultDelay = 1000, defaultTimeout = 20 }) => {
  const [timerCount, setTimerCount] = useState(0)
  const [delay, setDelay] = useState(null)
  const [handWashCompleted, setHandWashCompleted] = useState(false)

  const dispatch = useDispatch()

  const onUserPressButton = () => {
    setDelay(defaultDelay)
  }

  const onTimeOut = () => {
    setDelay(null)
    setHandWashCompleted(true)
  }

  const onHandWashCompleted = () => {
    dispatch(handWash.actions.addHandWash())
    setTimerCount(0)
    setHandWashCompleted(false)
  }

  useEffect(() => {
    if (timerCount === defaultTimeout) {
      onTimeOut()
    }
  }, [timerCount])

  useEffect(() => {
    if (handWashCompleted) {
      onHandWashCompleted()
    }
  }, [handWashCompleted])

  useInterval(() => {
    setTimerCount(timerCount + 1)
  }, delay)

  return (
    <View>
      <Text>Please press the button below to start the timer</Text>
      <Button
        onPress={onUserPressButton}
        testID="timer-button"
        title="Timer"
        disabled={timerCount > 0}
      />
      <Text testID="timer-counter">{timerCount}</Text>
    </View>
  )
}

HandWashComponent.propTypes = {
  defaultDelay: PropTypes.number,
  defaultTimeout: PropTypes.number,
}

export default HandWashComponent
