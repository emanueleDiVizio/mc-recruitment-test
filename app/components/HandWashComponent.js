import React, { useState, useEffect } from 'react'
import { View, Text, Button } from 'react-native'
import useInterval from '../hooks/useInterval'
import PropTypes from 'prop-types'

const HandWashComponent = ({ defaultDelay = 1000 }) => {
  const [timerCount, setTimerCount] = useState(0)
  const [delay, setDelay] = useState(null)
  const [handWashCompleted, setHandWashCompleted] = useState(false)

  const onUserPressButton = () => {
    setDelay(defaultDelay)
  }

  const onHandWashComplete = () => {
    setDelay(null)
    setHandWashCompleted(true)
  }

  useEffect(() => {
    if (timerCount === 20) {
      onHandWashComplete()
    }
  }, [timerCount])

  useInterval(() => {
    setTimerCount(timerCount + 1)
  }, delay)

  return (
    <View>
      <Text>Please press the button below to start the timer</Text>
      <Button onPress={onUserPressButton} testID="timer-button">
        Timer
      </Button>
      <Text testID="timer-counter">{timerCount}</Text>
    </View>
  )
}

HandWashComponent.propTypes = {
  defaultDelay: PropTypes.number,
}

export default HandWashComponent
