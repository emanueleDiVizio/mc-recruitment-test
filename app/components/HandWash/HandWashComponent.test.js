import React from 'react'
import HandWashComponent from './HandWashComponent'
import { fireEvent, render, act, wait } from '@testing-library/react-native'
import * as redux from 'react-redux'
import handWash from '../../state/handWashSlice'

jest.useFakeTimers()

describe('HandWashComponent', () => {
  let dispatch

  beforeEach(() => {
    dispatch = jest.fn()
    redux.useDispatch = () => dispatch
  })

  it('initializes correctly', () => {
    const { baseElement } = render(<HandWashComponent />)

    expect(baseElement).toMatchSnapshot()
  })

  it('starts the timer when button is pressed and updates with elapsed time', async () => {
    const { baseElement, getByTestId, getByText } = render(
      <HandWashComponent />
    )

    const button = getByTestId('timer-button')
    fireEvent.press(button)

    act(() => jest.advanceTimersByTime(10000))

    await wait(() => expect(getByText('10')).toBeTruthy())

    expect(baseElement).toMatchSnapshot()
  })

  it('dispatches `addHandWash` when timer is complete ', async () => {
    const { getByTestId } = render(<HandWashComponent />)

    const button = getByTestId('timer-button')
    fireEvent.press(button)

    act(() => jest.runAllTimers())

    expect(dispatch).toHaveBeenCalledWith(handWash.actions.addHandWash())
  })
})
