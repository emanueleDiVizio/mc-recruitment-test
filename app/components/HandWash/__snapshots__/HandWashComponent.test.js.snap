// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandWashComponent initializes correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text>
        Please press the button below to start the timer
      </Text>
      <Button
        disabled={false}
        testID="timer-button"
        title="Timer"
      />
      <Text
        testID="timer-counter"
      >
        0
      </Text>
    </View>
  </View>
</View>
`;

exports[`HandWashComponent starts the timer when button is pressed and updates with elapsed time 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text>
        Please press the button below to start the timer
      </Text>
      <Button
        disabled={true}
        testID="timer-button"
        title="Timer"
      />
      <Text
        testID="timer-counter"
      >
        10
      </Text>
    </View>
  </View>
</View>
`;
